### PR TITLE
nixos/steam: Add fonts to /usr/share/fonts in the FHS

### PR DESCRIPTION
**Pull request by AmandaCameron**
Some unity games don't use fontconfig for finding fonts to use on the system, instead they manually poll /usr/share/fonts and parse the files themselves. This patch makes it so that the fonts in NixOS are passed into the FHS.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/programs/steam.nix [hunk 1]
```diff
@@ -6,6 +6,14 @@ let
   cfg = config.programs.steam;
 
   steam = pkgs.steam.override {
+    extraPackages = pkgs: [
```
Comment (xaverdh) [inline nixos/modules/programs/steam.nix:9]: should this be `extraPkgs` instead?

Comment (xaverdh) [reply nixos/modules/programs/steam.nix:9]: (judging from https://github.com/NixOS/nixpkgs/blob/440405af3b5f52d35362c2b3501888d7c728e2b4/pkgs/games/steam/fhsenv.nix#L3)

Comment (AmandaCameron) [reply nixos/modules/programs/steam.nix]: Whoops, that's absolutely correct, fixed.

File: nixos/modules/programs/steam.nix [hunk 2]
```diff
@@ -6,6 +6,14 @@ let
   cfg = config.programs.steam;
 
   steam = pkgs.steam.override {
+    extraPkgs = pkgs: [
```
Comment (imsofi) [inline nixos/modules/programs/steam.nix:9]: I feel like this should be done inside of the `fhsenv.nix` file, and not through `extraPkgs`. As this option is used by people to override and add extra packages on their end. So this change would get overwritten when someone does `steam.override { extraPkgs = (pkgs: ...); }` themselves.

**Review by soupglasses — CHANGES_REQUESTED** [review]

File: nixos/modules/programs/steam.nix (hunk 2 above)
Comment (xaverdh) [reply nixos/modules/programs/steam.nix:9]: Well you could do something like

```nix
lib.makeOverridable
  ( { extraPkgs, ... }@args:
    steam.override (args // { extraPkgs = pkgs: [...] ++ extraPkgs pkgs; }) )
  { extraPkgs = pkgs: []; }
```
(by the way, someone more experienced should check if that is the correct way to do it).

But its probably easier to edit `fhsenv.nix` instead..

**Comment by AmandaCameron** [comment]
fhsenv.nix doesn't have access to the nixos configuration does it? This
passes through the nixos fonts.fonts option so that the user's selected
fonts can be accessed by the steam FHS when under NixOS

**Comment by xaverdh** [comment]
> fhsenv.nix doesn't have access to the nixos configuration does it? This passes through the nixos fonts.fonts option so that the user's selected fonts can be accessed by the steam FHS when under NixOS

Maybe just add a new optional argument for passing fonts to `fhsenv.nix` (like a list of paths to search for fonts), so the nixos module can inject them there?

**Comment by imsofi** [comment]
I think the best option here would be to not rely on `config.fonts.fonts`, as it is dependent on each system, rather than a reproducible list of fonts. Rather setting up a `/usr/share/fonts` inside of `fhsenv.nix` which contains the one or two specific fonts that the game requires, with a comment of `# Game Name` for its reasoning to be added.

File: nixos/modules/programs/steam.nix (hunk 2 above)
Comment (imsofi) [reply nixos/modules/programs/steam.nix:9]: This is still a really hacky way to include a font, and should rather be refactored into something more reproducible inside of the `fhsenv` itself.

**Comment by AmandaCameron** [comment]
Sorry for the delay, lack of spoons hit me recently. I went and added an extraFonts option to the FHS, which the nixos module adds the value of config.fonts.fonts to. This could also have a list of fonts added by the fhs itself for common fonts for some games, if anybody has any examples for this.

**Comment by Shawn8901** [comment]
i came around here by random and noticed that this PR fixes the font issues i had with some steam games for a longer time and no one could help me. 
As a "workaround" for personal flake its enough to have something like

```
let 
  fontsPkg = pkgs: (pkgs.runCommand "share-fonts" { preferLocalBuild = true; } ''
    mkdir -p "$out/share/fonts"
    font_regexp='.*\.\(ttf\|ttc\|otf\|pcf\|pfa\|pfb\|bdf\)\(\.gz\)?'
    find ${toString (config.fonts.fonts)} -regex "$font_regexp" \
      -exec ln -sf -t "$out/share/fonts" '{}' \;
  '');
in
{
  nixpkgs.config.packageOverrides = pkgs: {
    steam = pkgs.steam.override {
      extraPkgs = pkgs: with pkgs; [
        (fontsPkg pkgs)
      ];
    };
  };
}
```

But as its a manual step and getting the fonts into the FHS at the right place that would drastically improve the UX for unity games.
The game where i recognized that the most was Universim which was basically unplayable due half of the texts were missing.

Would be sad if the PR gets stale.

**Review by soupglasses — CHANGES_REQUESTED** [review]
Hi again! I see we are still using `config.fonts.fonts` for the PR, but that is problematic as its based off your nixos configuration and will not build in a pure context.

This is since `config.fonts.fonts` is only set if you build it together with a NixOS configuration whose enabled `config.fonts.enableDefaultFonts`, `config.services.xserver.enable`, or set up by yourself.[^1]

If there is a font missing, for example if liberation fonts were required for a game, adding `pkgs.liberation_ttf` into the packages could work. It would require testing and naming the game which that fixes. I am unsure if adding fonts into the fhs is best practice, but it is better than relying on `config.fonts.fonts`.

Also reading online about this issue, it may be an issue with some Unity games relying on non-free Microsoft fonts only?[^2] If so, installing `corefonts` through wine or winetricks could be the fix for these specific games. Again, testing which fonts a game needs, then naming the game that needed them is of much help with how we can go forwards with this PR.

[^1]: https://github.com/NixOS/nixpkgs/blob/2e751c0772b9d48ff6923569adfa661b030ab6a2/nixos/modules/config/fonts/fonts.nix#L76-L79
[^2]: https://www.reddit.com/r/wine_gaming/comments/tlp3st/issue_with_unity_games/

**Comment by xaverdh** [comment]
So I think we all agree that adding the `extraFonts` hooks to the `fhsenv` wrapper is a good change, right? The discussion is about what to put there. Maybe instead of hard coding `config.fonts.fonts` we can add a new option to the nixos module (which defaults to `config.fonts.fonts`), so people can decide themselves?
That should also keep the door open to making the list of fonts smaller in the future.

**Comment by xaverdh** [comment]
also pinging @jagajaga @jonringer since hey are listed as maintainers for steam

**Comment by xaverdh** [comment]
and @mkg20001 who is listed as maintainer of the nixos module

**Comment by AmandaCameron** [comment]
I'll be honest. I lack the social spoons to deal with further going back and forth on this PR. If someone else wants to take the work done so far, and make it acceptable, I'm more than welcome to let them.

The issue is, as I understand it, that some unity games (At the very least, the one that lead me to making this PR, Kerbal Space Program) don't use fontconfig or any other native linux function for looking up fonts. I decuded this because KSP was doing an opendir syscall to `/usr/share/fonts`, and fonts were missing when in-game. The solution for this for me was to make `/usr/share/fonts` exist inside the FHS, by way of a [patch to my local nixos flake](https://git.camnet.site/homelab/nixos/commit/2dd8f189685ef22ba9bbdae9b477d4da80316e6b) which made the relevent files exist inside the FHS, this resulted in fonts being rendered in-game, at least after adding them to my flake's relevent file, but is highly specific to my current flake setup, so I tried to generalise it for this PR.

**Pull request closed without merging**